### ISSUE TITLE
Feature request: Callback function gets passed current best solution

### ISSUE THREAD

**Issue by ngiann** [JuliaNLSolvers/Optim.jl#452]
Hello everyone,
I want to propose a feature for the callback function that gets called during tracing.

If i understand correctly, the callback function is passed an OptimizationState which contains information on the current iteration, current value and gradient norm.

Would it be possible to also pass the best solution found by the solver so far?

This would be useful for plotting/printing/reporting the current best solution while waiting for a lengthy optimisation to complete.

Thanks.

**Comment by ngiann** [comment]
Actually this would also be good when training models: a callback function, called at every or every other iterations, could be useful for monitoring training error on a hold out set.

**Comment by pkofod** [comment]
Sorry, missed your issue when you created it. Yes I am very open to giving access to more (all?) of the state.

**Comment by ngiann** [comment]
Hi pkofod, thanks for your response.

I tried looking into the actual code to see whether it is possible to access the best current solution encountered by the optimiser so far, but either it is indeed not recorded or it is there but I don't know how to retrieve it.

I don't know if other pieces of information should also be exposed. However, I do think that exposing the current best solution would be extremely useful. I think the example I gave, of monitoring overfitting on a hold out dataset, is something that many people would be interested in. Were this information to become available, maybe I could contribute to the documentation a somewhat lengthier example of how to use the Callback option to monitor the training of, let's say, a very simple neural network.

Cheers, N.

**Comment by pkofod** [comment]
Actually, you can already do this. I just forgot I fixed the `extended_trace` bug. Consider

```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0 (2017-06-19 13:05 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

julia> using Optim

julia> srand(813423)
MersenneTwister(UInt32[0x000c696f], Base.dSFMT.DSFMT_state(Int32[-1161004618, 1073592952, 381213596, 1073231558, -1931639110, 1073201536, -820244687, 1073046726, 910572206, 1073274658  …  -1534945524, 1073515202, -1453155325, 1072911297, 405483409, -2010084009, 1571460297, -2113672915, 382, 0]), [1.26361, 1.62117, 1.01768, 1.77406, 1.97267, 1.61661, 1.58569, 1.28964, 1.94363, 1.07836  …  1.01537, 1.54896, 1.04484, 1.20191, 1.45115, 1.17675, 1.29676, 1.81861, 1.8633, 1.09164], 382)

julia> x=rand(20,3)
20×3 Array{Float64,2}:
 0.203463   0.0301294   0.309577  
 0.623049   0.755929    0.00900303
 0.981069   0.895044    0.15804   
 0.514308   0.594124    0.585507  
 0.058079   0.578574    0.283453  
 0.311425   0.51559     0.693598  
 0.943308   0.0650927   0.419773  
 0.730943   0.127825    0.137685  
 0.332714   0.604295    0.565532  
 0.34213    0.92734     0.697526  
 0.699817   0.756698    0.593537  
 0.47088    0.849869    0.0163709 
 0.417274   0.648776    0.0166088 
 0.0725709  0.640063    0.408913  
 0.139407   0.00575509  0.972732  
 0.266509   0.399887    0.433635  
 0.782165   0.272634    0.718461  
 0.226823   0.155478    0.421809  
 0.475578   0.264313    0.21655   
 0.914196   0.789476    0.239771  

julia> y=x*[1;2;3]+randn(20)
20-element Array{Float64,1}:
 0.643799
 2.20267 
 3.38555 
 3.27363 
 1.7783  
 3.84733 
 0.688427
 2.23531 
 2.40242 
 3.67807 
 3.67385 
 0.961567
 2.29233 
 4.19935 
 2.96024 
 2.77309 
 2.93642 
 2.79894 
 1.50142 
 2.36746 

julia> function cb(os)
           # do something with the current coefficent vector, here we print it
           println("<callback>")
           println(os.metadata["x"])
           println("</callback>")
           # false means don't stop the optimization procedure
           # if you want the callback to stop, return true
           return false
       end
cb (generic function with 1 method)

julia> optimize(d->sum(z->z^2, x*d-y), [0.0;0.0;0.0], BFGS(), Optim.Options(extended_trace=true, callback=cb))
<callback>
[0.0, 0.0, 0.0]
</callback>
<callback>
[1.7004, 1.98146, 1.64106]
</callback>
<callback>
[0.473798, 2.03465, 3.15145]
</callback>
<callback>
[0.275896, 2.29777, 3.04671]
</callback>
Results of Optimization Algorithm
 * Algorithm: BFGS
 * Starting Point: [0.0,0.0,0.0]
 * Minimizer: [0.27589557408983034,2.2977683157012163, ...]
 * Minimum: 1.035707e+01
 * Iterations: 3
 * Convergence: true
   * |x - x'| < 1.0e-32: false 
     |x - x'| = 2.63e-01 
   * |f(x) - f(x')| / |f(x)| < 1.0e-32: false
     |f(x) - f(x')| / |f(x)| = NaN 
   * |g(x)| < 1.0e-08: true 
     |g(x)| = 2.89e-10 
   * stopped by an increasing objective: false
   * Reached Maximum Number of Iterations: false
 * Objective Calls: 7
 * Gradient Calls: 7

```

**Comment by pkofod** [comment]
Instead of printing, you can simply calculate the loss at the validation set

**Comment by ngiann** [comment]
This is really great.
So, os.metadata["x"] holds the current best solution then?
This must be definitely highlighted in the documentation, I will try to add it somehow.

**Comment by pkofod** [comment]
Yes `os` stands for `optimization state` in this context and `.metadata` is used to store extra data coming from `extended_trace` being `true`. The usual trace only holds things like norms (so scalars) because you don't accidentally want to store a full trace of `x` whenever you're optimizing something that iterates 5000 times with hundreds of thousands of variables (or whatever you problem is).

**Comment by ngiann** [comment]
Thanks for the last clarification.

**Comment by murphyk** [comment]
Storing the iterate `x` for all steps would indeed use a lot of memory for high dimensional problems.
It seems to me that you always want to pass `x` to the callback, even if `extended_trace=false`, so the history is not stored.
(Currently it seems the opt_state (for BFGS) just contains the following fields with metadata being an empty dict, unless extended_trace=true, in which case it contains these keys: `["Current step size", "g(x)", "~inv(H)", "x"]`.

**Comment by pkofod** [comment]
> So it seems that you should set extended_trace=true to pass metadata to the callback, but store_trace=false, to save memory in high dimensional settings.

Yes, storing the full trace with `extended_trace=true` can be... a lot.

This whole interface could do with a re-design. That is not a priority right now, though (I have limited time now, but I'm hoping to find support for more work in the future), but we could go a long way with some better documentation. It could be a "callback" section in each algorithm's docstring for example.

**Comment by ngiann** [comment]
It is really great that this feature was added. This very much appreciated!!!
